subsys/bluetooth/controller/ll_sw/lll_filter.h

```diff
@@ -25,3 +25,11 @@ extern bool ull_filter_lll_rl_addr_allowed(u8_t id_addr_type, u8_t *id_addr,
 extern bool ull_filter_lll_rl_addr_resolve(u8_t id_addr_type, u8_t *id_addr,
 					   u8_t rl_idx);
 extern bool ull_filter_lll_rl_enabled(void);
+#if defined(CONFIG_BT_CTLR_SW_DEFERRED_PRIVACY)
+typedef void (*ku_resolve_callback_t)(void *param);
+extern u8_t ull_filter_deferred_resolve(bt_addr_t *rpa,
```

Comment:
@joerchan and I are reviewing this and don't quite understand the need for the callback. Why not have something like:

```
/* Returns the rl_idx or triggers deferred resolution */
u8_t ull_filter_deferred_resolve(bt_addr_t *rpa);
```
and then have something in the LLL like:

```
lll.c:

rl_idx = ull_filter_deferred_resolve(rpa);
if (rl_idx != NONE) {
    /* use the resolved address */
} else {
    /* no resolution available, maybe next time we call ull_filter_deferred_resolve() it will be resolved */
}
```

Reply:
Initially a design without a callback was implemented, but this did not work due to long answer times on connection requests if I recall. A redesign was suggested by Peter:
1)	In case of new RPA reception, the LLL must defer addr check stuff and prepare Rx node with connection info as all address checks have passed (opposite order as in current Nordic LLL code).
2)	A pending resolving job flag is set (only one outstanding resolving job to limit deferred buffer allocation), and RPA resolving job is issued with callback function ptr as argument.
3)	When RPA resolving job is done, the callback function is enqueued for execution in LLL context and the pending resolving job flag is cleared.
4)	The callback function will do the address checking and forward the Rx node with connection info to ULL. In case of failing address check, the Rx node type is changed to none (node will just be released via ULL-LL path).

We have prepared a presentation on the topic (and will finalize with low level details based on actual LLL implementation) that is to be shared and discussed in the next ctrl f2f.

Reply:
Thanks for suggesting better names, I did implement suggested changes

Reply:
@asmk-ot I think it is quite difficult for us to review this seeing only one side here.

I imagined that once the resolving job was finished it notified the LLL about the resolution and then the LLL added it to a list of known resolutions. Meaning it could look up in the table. Which is why I suggested that maybe the table could be kept in ULL instead, that would avoid the problem of having to reimplement this for every LLL variant.

The design you outlined here you say "connection info", does that mean one instance of "connect to this guy if you see it"? Because if so I'm worried about the case where there are more than 1 privacy-enabled peer out there. Would you end up perpetually overwriting the connection info object and never being able to connect?

Reply:
I do realize that reviewing this is difficult when you only see half the implementation.
Let me try to get a the presentation together and I can share it with you guys. 
The implementation has a configurable size table for the known unknown peer RPA cache, in the current implementation there is a single work item that gets put on the work Q, that is a limitation. Once we have some more millage with the implementation we can decide if that needs to be redesigned to allow multiple outstanding resolve jobs, there is no limitation on the RX nodes beyond Q sizes.

Reply:
@asmk-ot Do you have any updates on the mentioned presentation? Could you share it with me also?

We are currently trying to build on this PR a LLL implementation for our platform (OpenISA). If we succeed, this might help you with the integration of the PR as it's full implementation would be available.

Reply:
@virgil-nxp yes let me know how I can reach you

Reply:
@asmk-ot  zephyr_ble@nxp.com is our internal mailing list. We can use it or talk on slack.
Is there something else that you prefer?

Reply:
Mail sent

Reply:
> The implementation has a configurable size table for the known unknown peer RPA cache, in the current implementation there is a single work item that gets put on the work Q, that is a limitation. Once we have some more millage with the implementation we can decide if that needs to be redesigned to allow multiple outstanding resolve jobs, there is no limitation on the RX nodes beyond Q sizes.

My question was not about the single work item in the work Q. It is about the storage of the resolved work item. 
Now after it has been resolved a function is called to provide some information to the LLL. And what happens with that is a black box.
The situation I'm asking about is this:
 - Adv PDU with RPA1 from Advertiser1 received.
 - Deferred resolve of RPA1 is started.
 - Deferred resolve of RPA1 is completed and LLL is notified
 - Adv PDU with RPA2 from Advertiser2 received
 - Deferred resolve of RPA2 is started.
 - Deferred resolve of RPA2 is completed and LLL is notified.

At this point what happens? Is the resolved address of Advertiser1 overwritten?
In that case you could end up in this pattern repeating itself and establishing the connection with either Advertiser1 or Advertiser2 might be delayed for an undefined amount of time.
Since I cannot tell what is happening I'm mainly asking if you have a solution for this problem, or there is no problem at all.

If you keep mulitple resolved entries in the LLL I'm wondering if there is perhaps better to have it in the ULL, there is already a peer_rpa entry in the resolving list.
LLL could then instead of having it's own cache, look at the entries in the resolving list for a matched resolved entry.
And the delayed resolved work would instead of notifying the LLL simply update the entry in the resolving list.
I imagine this approach would be easier to support across multiple variants in LLL as more of the feature is put into the ULL which is common.

Instead of having a cache of resolved items, which the LLL could look up during the T_IFS ISR.

Reply:
Addresses are kept in the ull_filtr.c rl (resolve list) structure for multiple sources, it is not overwritten, this has always been the case before we added the SW deferred privacy functionality. The "new" thing here is that the prpa_cache_resolve will either add the RPA to the "known unknown" cache if address cannot be resolved, or add it to the  resolve list under the index j, this is the index that corresponds to the IRK that was used to resolve the address.

Multiple entries are not kept in Oticons LLL implementation. As discussed at the F2F the way forward we be to merge these changes, Nordic (at others) can take a stab at implementing a solution (LLL part) and then we can discuss needed refactoring, that seems to be the most pragmatic way forward as it is a somewhat complicated topic. We have also shared an architecture document that sketches a possible LLL implementation with relevant parties.